File: dgs/src/test/java/com/example/demo/QueryTests.java
package com.example.demo;

import com.example.demo.gql.datafetchers.PostsDataFetcher;
import com.example.demo.gql.types.Post;
import com.example.demo.service.AuthorService;
import com.example.demo.service.PostNotFoundException;
import com.example.demo.service.PostService;
import com.netflix.graphql.dgs.DgsQueryExecutor;
import com.netflix.graphql.dgs.autoconfig.DgsAutoConfiguration;
import lombok.extern.slf4j.Slf4j;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.autoconfigure.ImportAutoConfiguration;
import org.springframework.boot.autoconfigure.jackson.JacksonAutoConfiguration;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.Import;

import java.util.List;
import java.util.Map;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.*;

@SpringBootTest(classes = {QueryTests.QueryTestsConfig.class})
@Slf4j
class QueryTests {

    @Autowired
    DgsQueryExecutor dgsQueryExecutor;

    @MockBean
    PostService postService;

    @MockBean
    AuthorService authorService;

    @Configuration
    @Import(PostsDataFetcher.class)
    @ImportAutoConfiguration(classes = {
            DgsAutoConfiguration.class,
            JacksonAutoConfiguration.class
    })
    static class QueryTestsConfig {

    }

    @Test
    void allPosts() {
        when(postService.getAllPosts())
                .thenReturn(
                        List.of(
                                Post.builder().id(1L)
                                        .title("test title")
                                        .content("test content")
                                        .build(),
                                Post.builder().id(2L)
                                        .title("test title2")
                                        .content("test content2")
                                        .build()
                        )
                );
        List<String> titles = dgsQueryExecutor.executeAndExtractJsonPath(
                " { allPosts { title content }}",
                "data.allPosts[*].title");

        assertThat(titles).containsAll(List.of("test title", "test title2"));

        verify(postService, times(1)).getAllPosts();
        verifyNoMoreInteractions(postService);
    }

    @Test
    void postById() {
        when(postService.getPostById(anyLong())).thenReturn(
                Post.builder().id(1L)
                        .title("test title")
                        .content("test content")
                        .build()
        );
        String title = dgsQueryExecutor.executeAndExtractJsonPath(
                "query postById($postId:ID!){ postById(postId:$postId) { title content }}",
                "data.postById.title",
                Map.of("postId", 1L)
        );

        assertThat(title).isEqualTo("test title");

        verify(postService, times(1)).getPostById(anyLong());
        verifyNoMoreInteractions(postService);
    }

    @Test
    void postById_notFound() {
        when(postService.getPostById(anyLong())).thenThrow(new PostNotFoundException(1L));
        var executeResult = dgsQueryExecutor.execute(
                "query postById($postId: ID!){ postById(postId:$postId) { title content }}",
                Map.of("postId", 1L)
        );


        assertThat(executeResult.getErrors()).isNotEmpty();
        assertThat(executeResult.getErrors().get(0).getMessage()).contains("Post: #1 was not found.");

        verify(postService, times(1)).getPostById(anyLong());
        verifyNoMoreInteractions(postService);
    }

}

File: dgs-kotlin-co/src/test/kotlin/com/example/demo/QueryTests.kt
package com.example.demo

import com.example.demo.gql.datafetcher.PostsDataFetcher
import com.example.demo.gql.scalar.LocalDateTimeScalar
import com.example.demo.gql.scalars.UUIDScalar
import com.example.demo.gql.types.Post
import com.example.demo.service.PostService
import com.netflix.graphql.dgs.reactive.DgsReactiveQueryExecutor
import com.netflix.graphql.dgs.test.EnableDgsTest
import com.ninjasquad.springmockk.MockkBean
import graphql.schema.GraphQLScalarType
import graphql.schema.idl.RuntimeWiring
import io.kotest.matchers.collections.shouldContainAll
import io.mockk.coEvery
import io.mockk.coVerify
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.flow.flowOf
import kotlinx.coroutines.reactor.awaitSingle
import kotlinx.coroutines.test.runTest
import org.junit.jupiter.api.Test
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.autoconfigure.ImportAutoConfiguration
import org.springframework.boot.autoconfigure.web.reactive.WebFluxAutoConfiguration
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.context.annotation.Import
import org.springframework.graphql.execution.RuntimeWiringConfigurer
import java.util.*

@OptIn(ExperimentalCoroutinesApi::class)
@SpringBootTest(classes = [QueryTests.TestConfig::class])
@EnableDgsTest
class QueryTests {
    companion object {
        private val log = LoggerFactory.getLogger(QueryTests::class.java)
    }

    @Configuration
    @Import(
        value = [
            PostsDataFetcher::class,
            LocalDateTimeScalar::class
        ]
    )
    @ImportAutoConfiguration(
        value = [
            WebFluxAutoConfiguration::class
        ]
    )
    class TestConfig{
        @Bean
        fun customRuntimeWiring(): RuntimeWiringConfigurer{
            return object: RuntimeWiringConfigurer {
                override fun configure(builder: RuntimeWiring.Builder) {
                    builder.scalar(
                        GraphQLScalarType.newScalar()
                            .name("UUID")
                            .description("UUID type")
                            .coercing(UUIDScalar())
                            .build()
                    )
                }
            }
        }
    }

    @Autowired
    lateinit var dgsQueryExecutor: DgsReactiveQueryExecutor

    @MockkBean
    lateinit var postService: PostService

    @Test
    fun `get all posts`() = runTest {
        coEvery { postService.allPosts() } returns
                flowOf(
                    Post(
                        id = UUID.randomUUID(),
                        title = "test title",
                        content = "test content"
                    ),
                    Post(
                        id = UUID.randomUUID(),
                        title = "test title2",
                        content = "test content2"
                    ),
                )

        val query = """
            query allPosts 
            { 
                allPosts 
                { 
                    title 
                    content 
                }
            }
        """.trimIndent()

        val titles = dgsQueryExecutor
            .executeAndExtractJsonPath<List<String>>(
                query,
                "data.allPosts[*].title"
            )
            .awaitSingle()

        titles shouldContainAll listOf("test title", "test title2")

        //verify
        coVerify(exactly = 1) { postService.allPosts() }
    }
}